### PR TITLE
Add tests for MultiSigWallet contract

### PR DESCRIPTION
**Pull request by harzo**

### CODE REVIEW

**Comment by jstefanski** [comment]
@harzo If PR is ready to review remove the `[WIP]` prefix.

**Comment by harzo** [comment]
Sorry for that :)
Done

File: test/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    0,
+    await getData(method, ...args),
+    { from }
+  );
+  assert.isOk(tx);
+
+  const transactionId = findLastTransactionId(tx);
+  for (let i = 1; i < owners.length; i++) {
+    tx = await wallet.confirmTransaction(transactionId, { from: owners[i] });
+    assert.isOk(tx);
```
Comment (jstefanski) [inline test/helpers.ts]: `assert.isOk()` checks only whether parameter is boolean true (not null etc.). I don't see the point in checking it's null, as it is never expected.

File: test/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  }
+  return tx;
```
Comment (jstefanski) [inline test/helpers.ts]: There are several transactions, why do you return only the last one?

File: types/wallet.d.ts [hunk 3]
```diff
@@ -21,13 +22,121 @@ declare module 'wallet' {
...
+        transactionId: AnyNumber,
+        options?: TransactionOptions
+      ): Promise<TransactionResult>;
+
+      executeTransaction(
+        transactionId: AnyNumber,
+        options?: TransactionOptions
+      ): Promise<TransactionResult>;
+
+      getConfirmationCount(transactionId: AnyNumber): Promise<AnyNumber>;
```
Comment (jstefanski) [inline types/wallet.d.ts]: Return value should be `Promise<BigNumber>`. Web3 takes as a parameter any of string, number, and BigNumber but returns always BigNumber.

File: types/wallet.d.ts [hunk 4]
```diff
@@ -21,13 +22,121 @@ declare module 'wallet' {
       ): Promise<TransactionResult>;
     }
 
+    interface MultiSigWallet extends ContractBase {
+      owners(): Promise<Address[]>;
+      required(): AnyNumber;
```
Comment (jstefanski) [inline types/wallet.d.ts]: Return value should be `Promise<BigNumber>`

File: types/wallet.d.ts [hunk 5]
```diff
@@ -21,13 +22,121 @@ declare module 'wallet' {
...
+        transactionId: AnyNumber,
+        options?: TransactionOptions
+      ): Promise<TransactionResult>;
+
+      getConfirmationCount(transactionId: AnyNumber): Promise<AnyNumber>;
+
+      getTransactionCount(
+        pending: boolean,
+        executed: boolean
+      ): Promise<AnyNumber>;
```
Comment (jstefanski) [inline types/wallet.d.ts]: Return value should be `Promise<BigNumber>`

File: types/wallet.d.ts [hunk 6]
```diff
@@ -21,13 +22,121 @@ declare module 'wallet' {
       ): Promise<TransactionResult>;
     }
 
+    interface MultiSigWallet extends ContractBase {
+      owners(): Promise<Address[]>;
```
Comment (jstefanski) [inline types/wallet.d.ts]: The correct autogenerated getter looks like `owners(num: AnyNumber): Promise<Address>`

File: test/multisigwallet.test.ts [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+} from './helpers';
+
+declare const web3: Web3;
+declare const artifacts: WalletArtifacts;
+declare const contract: ContractContextDefinition;
+
+const MultiSigWalletContract = artifacts.require('MultiSigWallet');
+
+contract('MultiSigWallet', accounts => {
+  const owners = [accounts[0], accounts[1], accounts[2]];
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: use `accounts.slice(...)`

File: test/multisigwallet.test.ts [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+  const required = 3;
+
+  let wallet: MultiSigWallet;
+
+  beforeEach(async () => {
+    wallet = await MultiSigWalletContract.new(owners, required, {
+      from: owners[0]
+    });
+    assert.isOk(wallet);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Remove, checks nothing.

File: test/multisigwallet.test.ts [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+  });
+
+  describe('#ctor', () => {
+    it('should revert for empty list of owners', async () => {
+      await assertReverts(async () => {
+        await MultiSigWalletContract.new([], 0);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Will revert because of `requirement = 0`, test is not isolated.

File: test/multisigwallet.test.ts [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+    });
+
+    it('should revert for zero required', async () => {
+      await assertReverts(async () => {
+        await MultiSigWalletContract.new(owners, 0);
+      });
+    });
+
+    it('should create wallet', async () => {
+      assert.ok(wallet);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Remove this testcase, tests nothing.

File: test/multisigwallet.test.ts [hunk 11]
```diff
@@ -0,0 +1,581 @@
...
+    it('should not have any transactions', async () => {
+      assert.equal(await wallet.getTransactionCount(true, true), 0);
+    });
+  });
+
+  describe('#submitTransaction', () => {
+    let tx: TransactionResult;
+
+    context('with invalid data', async () => {
+      it('should revert for non-existing owner', async () => {
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Replace `non-existing owner` with `non-owner`.

Comment (jstefanski) [reply test/multisigwallet.test.ts]: Futhermore, the test case is not isolated. It has two reasons to fail. First - called by non-owner, Second - invalid data.

File: test/multisigwallet.test.ts [hunk 12]
```diff
@@ -0,0 +1,581 @@
...
+          from: owners[0]
+        });
+      });
+
+      it('should emit SubmissionEvent', async () => {
+        const log = findLastLog(tx, 'Submission');
+        assert.isOk(log);
+
+        const event = log.args as SubmissionEvent;
+        assert.isOk(event);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Test event parameters

File: test/multisigwallet.test.ts [hunk 13]
```diff
@@ -0,0 +1,581 @@
...
+          wallet.address,
+          0,
+          await getData(wallet.addOwner, accounts[9]),
+          { from: owners[0] }
+        )
+      );
+      tx = await wallet.confirmTransaction(transactionId, { from: owners[1] });
+    });
+
+    context('with invalid data', async () => {
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: These reverts will always happen due to invalid data. Isolate tests.

File: test/multisigwallet.test.ts [hunk 14]
```diff
@@ -0,0 +1,581 @@
...
+        );
+      });
+
+      it('should revert for not confirmed transaction', async () => {
+        await assertReverts(async () => {
+          await wallet.executeTransaction(transactionId, { from: owners[1] });
+        });
+      });
+
+      beforeEach(async () => {
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: There are two `beforeEach` blocks. Merge them.

File: test/multisigwallet.test.ts [hunk 15]
```diff
@@ -0,0 +1,581 @@
...
+
+        const event = log.args as ExecutionEvent;
+        assert.isOk(event);
+      });
+
+      it('should set transaction as executed', async () => {
+        const transactionCount = await wallet.getTransactionCount(false, true);
+        assert.equal(transactionCount, 1);
+
+        const transactionIds = await wallet.getTransactionIds(
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Create helper functions like `getPendingTransactionIds` etc.

**Comment by harzo** [comment]
@jstefanski Tests fixed and refactored. Please, review again  👍

File: test/multisigwallet.test.ts [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+      const log = findLastLog(tx, 'RequirementChange');
+      assert.isOk(log);
+      const event = log.args as RequirementChangeEvent;
+      assert.isOk(event);
+    });
+
+    it('should change `required` attribute', async () => {
+      const tx = await executeChangeRequirements(new BigNumber(newRequired));
+
+      const walletRequired = findNewRequirement(tx);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: You should check `required` against getter not event.

Comment (jstefanski) [reply test/multisigwallet.test.ts]: This way you duplicate the previous test case 'should emit RequirementChange event'

File: test/multisigwallet.test.ts [hunk 17]
```diff
@@ -0,0 +1,521 @@
...
+
+      tx = await executeChangeRequirements(new BigNumber(4));
+
+      log = findLastLog(tx, 'ExecutionFailure');
+      assert.isOk(log);
+      event = log.args;
+      assert.isOk(event);
+    });
+
+    it('should emit RequirementChangeEvent', async () => {
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Replace with 'should emit RequirementChange event'

File: test/multisigwallet.test.ts [hunk 18]
```diff
@@ -0,0 +1,521 @@
...
+      assert.isOk(log);
+      event = log.args;
+      assert.isOk(event);
+
+      tx = await executeChangeRequirements(new BigNumber(4));
+
+      log = findLastLog(tx, 'ExecutionFailure');
+      assert.isOk(log);
+      event = log.args;
+      assert.isOk(event);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Check event parameters

File: test/multisigwallet.test.ts [hunk 19]
```diff
@@ -0,0 +1,521 @@
...
+      });
+    });
+
+    it('should emit ExecutionFailureEvent for non-owner', async () => {
+      const tx = await executeReplaceOwner(newOwner, newOwner);
+
+      const log = findLastLog(tx, 'ExecutionFailure');
+      assert.isOk(log);
+      const event = log.args as ExecutionFailureEvent;
+      assert.isOk(event);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Check event parameters

File: test/multisigwallet.test.ts [hunk 20]
```diff
@@ -0,0 +1,521 @@
...
+    });
+
+    it('should emit ExecutionFailureEvent for existing new owner', async () => {
+      const tx = await executeReplaceOwner(oldOwner, oldOwner);
+
+      const log = findLastLog(tx, 'ExecutionFailure');
+      assert.isOk(log);
+      const event = log.args as ExecutionFailureEvent;
+      assert.isOk(event);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Check event parameters

**Comment by jstefanski** [comment]
@harzo Good work, I like your tests very much. However, there are still some issues.
- event parameters are not tested
- some test cases use events instead of getters to check state of the contract (these are completely separate), you have no guarantee that storage contains the same data as event
- order of tests cases is not preferred, start with positive test cases, and keep `reverts` tests at the end

File: test/helpers.ts [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+
+  return new BigNumber(transactionId);
+}
+
+export function findNewRequirement(tx: TransactionResult) {
+  const log = findLastLog(tx, 'RequirementChange');
+  assert.isOk(log);
+
+  const paramName = 'required';
+  const newRequired = log.args[paramName] as Web3.AnyNumber;
```
Comment (jstefanski) [inline test/helpers.ts]: Why don't you use typed interface of event? I think the whole `findNewRequirment` function is redundant. Is is used only once.

File: test/helpers.ts [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+  const [param] = request.params;
+  return param.data;
+}
+
+export function findLastTransactionId(tx: TransactionResult) {
+  const log = findLastLog(tx, 'Submission');
+  assert.isOk(log);
+
+  const paramName = 'transactionId';
+  const transactionId = log.args[paramName] as Web3.AnyNumber;
```
Comment (jstefanski) [inline test/helpers.ts]: Use typed interface of event

File: test/helpers.ts [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+  assert.isOk(transactionId);
+
+  return new BigNumber(transactionId);
```
Comment (jstefanski) [inline test/helpers.ts]: No need to cast it to BigNumber. All uint256 numbers returned from Web3 are BigNumbers.

**Comment by harzo** [comment]
@jstefanski Again, tests fixed and refactored. Please, review 👍

File: test/multisigwallet.test.ts [hunk 24]
```diff
@@ -0,0 +1,519 @@
...
+      let log: TransactionLog;
+      let event: ExecutionFailureEvent;
+      result = await executeChangeRequirements(new BigNumber(0));
+      log = findLastLog(result.lastTransaction, 'ExecutionFailure');
+      assert.isOk(log);
+
+      event = log.args;
+      assertNumberEqual(result.transactionId, event.transactionId);
+
+      result = await executeChangeRequirements(new BigNumber(4));
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: A test case should test only a single thing. You test 0 and 4 requirements in one tests.

File: test/multisigwallet.test.ts [hunk 25]
```diff
@@ -0,0 +1,519 @@
...
+    });
+  });
+
+  describe('#changeRequirement', () => {
+    const newRequired = new BigNumber(2);
+
+    it('should emit ExecutionFailureEvent for invalid requirements', async () => {
+      let result: ExecutionResult;
+      let log: TransactionLog;
+      let event: ExecutionFailureEvent;
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: Avoid `let`. Use `const` wherever possible. When you split this test case into a proper one, you won't need reusing the variables. Always try to make things immutable in a functional programming fashion.

File: test/multisigwallet.test.ts [hunk 26]
```diff
@@ -0,0 +1,519 @@
...
+      log = findLastLog(result.lastTransaction, 'ExecutionFailure');
+      assert.isOk(log);
+
+      event = log.args;
+      assertNumberEqual(result.transactionId, event.transactionId);
+    });
+
+    it('should emit RequirementChangeEvent', async () => {
+      const result = await executeChangeRequirements(
+        new BigNumber(newRequired)
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: You don't need construction of new BigNumber

File: test/multisigwallet.test.ts [hunk 27]
```diff
@@ -0,0 +1,519 @@
...
+      assertNumberEqual(result.transactionId, event.transactionId);
+    });
+
+    it('should emit OwnerRemovalEvent', async () => {
+      const result = await executeReplaceOwner(oldOwner, newOwner);
+      const log = findLastLog(result.lastTransaction, 'OwnerRemoval');
+      assert.isOk(log);
+
+      const event = log.args as OwnerRemovalEvent;
+      assert.isOk(event);
```
Comment (jstefanski) [inline test/multisigwallet.test.ts]: You still don't test event parameters. Review all test cases for this issue.

**Comment by harzo** [comment]
@jstefanski Fixed, all isOk(event) changed to event parameters assertions, BigNumber removed from multisigwallet.test.ts file.

**Pull request merged**